linera-execution/src/transaction_tracker.rs

```diff
@@ -25,15 +25,36 @@ pub struct TransactionTracker {
     #[debug(skip_if = Vec::is_empty)]
     outcomes: Vec<ExecutionOutcome>,
     next_message_index: u32,
+    /// Events recorded by contracts' `emit` calls.
+    events: Vec<EventRecord>,
+    /// Subscribe chains to channels.
+    subscribe: Vec<(ChannelFullName, ChainId)>,
+    /// Unsubscribe chains from channels.
+    unsubscribe: Vec<(ChannelFullName, ChainId)>,
+}
+
+/// The [`TransactionTracker`] contents after a transaction has finished.
+#[derive(Debug, Default)]
+pub struct TransactionOutcome {
+    #[debug(skip_if = Vec::is_empty)]
+    pub oracle_responses: Vec<OracleResponse>,
+    #[debug(skip_if = Vec::is_empty)]
+    pub outcomes: Vec<ExecutionOutcome>,
+    pub next_message_index: u32,
+    /// Events recorded by contracts' `emit` calls.
+    pub events: Vec<EventRecord>,
+    /// Subscribe chains to channels.
+    pub subscribe: Vec<(ChannelFullName, ChainId)>,
+    /// Unsubscribe chains from channels.
+    pub unsubscribe: Vec<(ChannelFullName, ChainId)>,
```

Comment:
What happens if there are multiple actions to subscribe/unsubscribe for the same channel<>chainId? Will we try to "cancel them out" before handling?

Reply:
No, we just handle all of them.

These will go away anyway when we finally have event streams and remove pub-sub channels.

Reply:
It's wasteful to handle if they will cancel out each other... If you say we'll remove them then it doesn't matter though.